In [1]:
# Load packages
from costs_benefits_ssp.cb_calculate import CostBenefits
import pandas as pd
import os
import pathlib
import numpy as np


In [2]:
# Define paths
SSP_PATH = pathlib.Path(os.getcwd()).parent.parent.as_posix()
SSP_RUN  = os.path.join(SSP_PATH, "ssp_run_output/sisepuede_summary_results_run_sisepuede_run_2026-04-01T14;29;26.355576/nwr")


CB_DEFAULT_DEFINITION_PATH = os.path.join(SSP_PATH,"cost-benefits/cb_cost_factors")
CB_DEFAULT_DEFINITION_FILE_PATH = os.path.join(CB_DEFAULT_DEFINITION_PATH, "cb_config_params.xlsx")

CB_OUTPUT = os.path.join(SSP_PATH, "Tableau/data")

In [3]:
CB_OUTPUT

'/Users/alexa/Projects/ssp_uganda_data/ssp_modeling/Tableau/data'

In [4]:
CB_DEFAULT_DEFINITION_PATH

'/Users/alexa/Projects/ssp_uganda_data/ssp_modeling/cost-benefits/cb_cost_factors'

In [5]:
# Load data
ssp_data = pd.read_csv(os.path.join(SSP_RUN, "uganda.csv"))
att_primary = pd.read_csv(os.path.join(SSP_RUN,"ATTRIBUTE_PRIMARY.csv"))
att_strategy = pd.read_csv(os.path.join(SSP_RUN,"ATTRIBUTE_STRATEGY.csv"))
strategy_code_base = "BASE"

In [6]:
att_primary

,primary_id,design_id,strategy_id,future_id
0,0,0,0,0
1,700070,0,6004,0
2,760076,0,6010,0
3,770077,0,6011,0
4,780078,0,6012,0


In [7]:
# Instantiate CostBenefits object
cb = CostBenefits(ssp_data, att_primary, att_strategy, strategy_code_base)

# The export_db_to_excel method saves the initial configuration of the cost tables to an excel file.
# Each sheet represents a table in the cost and benefit program database.
# If the Excel file name is not given, the file will be saved with the default name cb_config_params.xlsx on the current python session.

#cb.export_db_to_excel(CB_DEFAULT_DEFINITION_FILE_PATH)

# Once that the excel file has been updated, we can reload it in order to update the cost factors database
cb.load_cb_parameters(CB_DEFAULT_DEFINITION_FILE_PATH)

The TX TX:FRST:INCREASE_SEQUESTRATION_NZ is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NZ is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NZ is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NZ is missing on AttTransformationCode
The TX TX:FRST:INCREASE_SEQUESTRATION_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NDC_25 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_25 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:DEC_WETLAND_LOSS_NDC_25 is missing on AttTransformationCode
The TX TX:LNDU:SET_WETLANDS_MINIMUM_NDC_25 is missing on AttTransformationCode
The TX TX:SCOE:INC_EFFICIENCY_HEAT_NDC_25 is missing on AttTransformationCode
Cargamos configuración de archivo excel
Se actualizó la base de datos


In [8]:
# Compute System Costs
results_system = cb.compute_system_cost_for_all_strategies()


************************************
*Strategy : PFLO:HBLE (0/4)
************************************

---------Costs for: cb:wali:technical_cost:sanitation:unimp_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:imp_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:safeman_rural.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:unimp_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:imp_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:safeman_urban.
La variable se evalúa en System Cost
---------Costs for: cb:wali:technical_cost:sanitation:omit_rural.
La variable se evalúa en System Cost
---------Costs for: cb:entc:technical_cost:electricity:capex.
La variable se evalúa en System Cost
---------Costs for: cb:entc:technical_cost:electricity:transmission.
L

In [9]:
# Compute Technical Costs
results_tx = cb.compute_technical_cost_for_all_strategies()


************************************
*Strategy : PFLO:HBLE (0/4)
************************************

---------Costs for: cb:inen:technical_cost:fuel_switch:lo_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:lo_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:fuel_switch:hi_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:hi_heat.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_savings:fuel_switch:all.
La variable se evalúa en Transformation Cost
---------Costs for: cb:inen:technical_cost:efficiency:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:technical_cost:electrification:LDV.
La variable se evalúa en Transformation Cost
---------Costs for: cb:trns:te

/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:856: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp["difference_variable"] = cb_orm.diff_var
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:857: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  tmp["difference_value"] = data_merged["difference"]
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:858: SettingWith

---------Costs for: cb:scoe:technical_savings:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_cost:efficiency:building_shell.
La variable se evalúa en Transformation Cost
---------Costs for: cb:entc:technical_cost:loss_reduction:electricity.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ippu:technical_cost:clinker:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ippu:technical_cost:abating_gases:n2o.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ippu:technical_cost:abating_gases:pfcs.
La variable se evalúa en Transformation Cost
---------Costs for: cb:ippu:technical_cost:abating_gases:other_fgases.
La variable se evalúa en Transformation Cost
---------Costs for: cb:fgtv:technical_cost:flaring:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:fgtv:technical_cost:leaks:X.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:

/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1067: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_baseline'] = 0
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1070: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_pathway'] = data_merged["difference_value"]



************************************
*Strategy : PFLO:NDC_25_UNCONDITIONAL (2/4)
************************************

---------Costs for: cb:scoe:technical_cost:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_savings:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_cost:efficiency:building_shell.
La variable se evalúa en Transformation Cost
---------Costs for: cb:waso:technical_cost:biogas_capture.
La variable se evalúa en Transformation Cost

************************************
*Strategy : PFLO:NDC_25_UNCONDITIONAL_CE (3/4)
************************************

---------Costs for: cb:scoe:technical_cost:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_savings:fuel_switch:heat_pumps.
La variable se evalúa en Transformation Cost
---------Costs for: cb:scoe:technical_cost:efficiency:building_shell.
La variable se

/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1067: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_baseline'] = 0
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:1070: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  data_output['variable_value_pathway'] = data_merged["difference_value"]


In [10]:
# Combine results
results_all = pd.concat([results_system, results_tx], ignore_index = True)

#-------------POST PROCESS SIMULATION RESULTS---------------
# Post process interactions among strategies that affect the same variables
results_all_pp = cb.cb_process_interactions(results_all)

# SHIFT any stray costs incurred from 2015 to 2025 to 2025 and 2035
results_all_pp_shifted = cb.cb_shift_costs(results_all_pp)

Resolving Interactions in SCOE : TX:SCOE:INC_EFFICIENCY_APPLIANCE, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN : TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:SHIFT_FUEL_HEAT, TX:INEN:INC_EFFICIENCY_ENERGY 
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Interactions in INEN :  
No interactions, skipping... PFLO:NDC_25_UNCONDITIONAL
Resolving Interactions in SCOE : TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:SHIFT_FUEL_HEAT, TX:SCOE:DEC_DEMAND_HEAT 
Resolving Intera

/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:660: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  res_pre2025["variable"] = res_pre2025["variable"] + "_shifted" + (res_pre2025["time_period"]+SSP_GLOBAL_TIME_PERIOD_0).astype(str)#create a new variable so they can be recognized as shifted costs
/opt/anaconda3/envs/ssp_uganda_env/lib/python3.11/site-packages/costs_benefits_ssp/cb_calculate.py:661: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  res_pre2025["tim

In [11]:
results_all_pp_shifted.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0


In [12]:
ssp_data.head()

,primary_id,region,time_period,area_agrc_crops_bevs_and_spices,area_agrc_crops_cereals,area_agrc_crops_fibers,area_agrc_crops_fruits,area_agrc_crops_herbs_and_other_perennial_crops,area_agrc_crops_nuts,area_agrc_crops_other_annual,...,yf_agrc_herbs_and_other_perennial_crops_tonne_ha,yf_agrc_nuts_tonne_ha,yf_agrc_other_annual_tonne_ha,yf_agrc_other_woody_perennial_tonne_ha,yf_agrc_pulses_tonne_ha,yf_agrc_rice_tonne_ha,yf_agrc_sugar_cane_tonne_ha,yf_agrc_tubers_tonne_ha,yf_agrc_vegetables_and_vines_tonne_ha,yf_lndu_supremum_pastures_tonne_per_ha
0,0,uganda,4,634960.593061,2.025945e+06,156073.139455,1.131815e+06,505.429083,513832.892470,2.915383e+06,...,2.172065,0.458120,1.528325,0.3674,0.799814,2.970174,77.726965,4.805888,4.035314,92.81
1,0,uganda,5,640126.452982,2.042427e+06,157342.906406,1.141024e+06,509.541112,518013.291655,2.939102e+06,...,2.196794,0.611231,1.681739,0.3674,0.805568,3.032345,78.023706,4.440105,4.013004,92.81
2,0,uganda,6,642767.262368,2.050853e+06,157992.016628,1.145731e+06,511.643198,520150.329354,2.951227e+06,...,2.210866,0.841393,2.293907,0.3674,0.807398,3.080626,61.430414,4.911305,4.023409,92.81
3,0,uganda,7,645322.479649,2.059006e+06,158620.088334,1.150285e+06,513.677153,522218.102852,2.962959e+06,...,2.225270,0.400800,2.699337,0.3674,0.810044,3.094956,77.255948,4.924276,4.030111,92.81
4,0,uganda,8,647798.760608,2.066907e+06,159228.757514,1.154699e+06,515.648274,524221.998246,2.974329e+06,...,2.225270,0.400800,2.699337,0.3674,0.810044,3.094956,77.255948,4.924276,4.030111,92.81


# Indoor air pollution benef

Health co-benefit from reduced household air pollution

To monetize the health benefits of reducing indoor air pollution from the use of traditional biomass stoves, we drew on estimates from the BAR-HAP model (The Benefits of Action to Reduce Household Air Pollution; WHO, 2021).
The study reports cost-effectiveness ratios ranging from 674 – 869 USD per DALY avoided for clean cooking interventions in Nepal, meaning that each avoided disability-adjusted life year (DALY) costs approximately 700 USD.

To adapt this metric to Uganda, we scaled the cost-effectiveness ratio by the relative difference in GDP per capita between Nepal (~1,450 USD) and Uganda (~1,070 USD), yielding an adjusted range of ≈ 500–640 USD per DALY avoided.
Assuming that the transition from solid fuels to clean cooking solutions avoids about 0.1 DALY per person per year, the implied annual health benefit equals 50–64 USD per person.

Given an average household size of five, this corresponds to an estimated 250–320 USD per household per year in health-related economic benefits.

This adjusted value was used in the cost–benefit analysis as the annual household health co-benefit of eliminating exposure to indoor air pollution from firewood and other solid fuels in Uganda.

In [13]:
import pandas as pd
import numpy as np

# parameters
benefit_per_person = 320   # USD/person/year. 
years = np.arange(4, 56)

# ['primary_id', 'time_period', 'population_gnrl_total']

# baseline
baseline = ssp_data[ssp_data['primary_id'] == 0][['time_period', 'population_gnrl_total', 'frac_scoe_heat_energy_residential_solid_biomass']]
baseline = baseline.rename(columns={'population_gnrl_total': 'population_baseline'})
baseline = baseline.rename(columns={'frac_scoe_heat_energy_residential_solid_biomass': 'frac_scoe_heat_energy_residential_solid_biomass_baseline'})

baseline.head()

,time_period,population_baseline,frac_scoe_heat_energy_residential_solid_biomass_baseline
0,4,39853824,0.960324
1,5,40996690,0.932779
2,6,42172327,0.901015
3,7,43381679,0.901015
4,8,44625710,0.901015


In [14]:
# baseline merge
merged = ssp_data[ssp_data['primary_id'] != 0][['primary_id', 'time_period', 'population_gnrl_total', 
                                                'frac_scoe_heat_energy_residential_solid_biomass']].merge(baseline, 
                                                                                                          on='time_period', how='left')
merged.tail(50)

,primary_id,time_period,population_gnrl_total,frac_scoe_heat_energy_residential_solid_biomass,population_baseline,frac_scoe_heat_energy_residential_solid_biomass_baseline
158,780078,6,42172327,0.901015,42172327,0.901015
159,780078,7,43381679,0.901015,43381679,0.901015
160,780078,8,44625710,0.901015,44625710,0.901015
161,780078,9,45905416,0.901015,45905416,0.901015
162,780078,10,47221819,0.901015,47221819,0.901015
163,780078,11,48562777,0.901015,48562777,0.901015
164,780078,12,49928247,0.891400,49928247,0.892538
165,780078,13,51318160,0.876952,51318160,0.879923
166,780078,14,52732427,0.855544,52732427,0.861368
167,780078,15,54170936,0.824507,54170936,0.834544


In [15]:
# frac_scoe_heat_energy_residential_solid_biomass 
merged['pop_diff'] = (merged['frac_scoe_heat_energy_residential_solid_biomass_baseline'] - merged['frac_scoe_heat_energy_residential_solid_biomass'] ) * merged['population_gnrl_total']

In [16]:
# difference calculation
merged['value'] =  merged['pop_diff'] * benefit_per_person

In [17]:
merged.head(30)

,primary_id,time_period,population_gnrl_total,frac_scoe_heat_energy_residential_solid_biomass,population_baseline,frac_scoe_heat_energy_residential_solid_biomass_baseline,pop_diff,value
0,700070,4,39853824,0.960324,39853824,0.960324,0.000000e+00,0.000000e+00
1,700070,5,40996690,0.932779,40996690,0.932779,0.000000e+00,0.000000e+00
2,700070,6,42172327,0.901015,42172327,0.901015,0.000000e+00,0.000000e+00
3,700070,7,43381679,0.901015,43381679,0.901015,0.000000e+00,0.000000e+00
4,700070,8,44625710,0.901015,44625710,0.901015,0.000000e+00,0.000000e+00
5,700070,9,45905416,0.901015,45905416,0.901015,0.000000e+00,0.000000e+00
6,700070,10,47221819,0.901015,47221819,0.901015,0.000000e+00,0.000000e+00
7,700070,11,48562777,0.901015,48562777,0.901015,0.000000e+00,0.000000e+00
8,700070,12,49928247,0.878908,49928247,0.892538,6.805106e+05,2.177634e+08
9,700070,13,51318160,0.851210,51318160,0.879923,1.473473e+06,4.715115e+08


In [18]:
merged = merged.merge(att_primary[['primary_id', 'strategy_id', 'future_id']], on='primary_id', how='left')
merged = merged.merge(att_strategy[['strategy_id', 'strategy_code']], on='strategy_id', how='left')

In [19]:
#merged.to_csv(os.path.join(CB_OUTPUT, "cba_indoor_air_pollution.csv"), index = False)

In [20]:
merged['region']='uganda'
merged['difference_variable'] ='frac_scoe_heat_energy_residential_solid_biomass'
merged['difference_value'] = 'pop_diff'
merged['variable'] ='cb:scoe:indoor_air_pollution'

In [21]:
merged = merged[['strategy_code', 'future_id', 'region', 'time_period', 'difference_variable', 'value','variable']]
merged.head()

,strategy_code,future_id,region,time_period,difference_variable,value,variable
0,PFLO:HBLE,0,uganda,4,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution
1,PFLO:HBLE,0,uganda,5,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution
2,PFLO:HBLE,0,uganda,6,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution
3,PFLO:HBLE,0,uganda,7,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution
4,PFLO:HBLE,0,uganda,8,frac_scoe_heat_energy_residential_solid_biomass,0.0,cb:scoe:indoor_air_pollution


In [22]:
#merged.to_csv(os.path.join(CB_OUTPUT, "cba_indoor_air_pollution.csv"), index = False)

In [23]:
results_all_pp_shifted= pd.concat([results_all_pp_shifted, merged], ignore_index = True)

In [24]:
results_all_pp_shifted.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0


In [25]:
results_all_pp_shifted.tail()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value
50029,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,51,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.229103e+07
50030,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,52,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.275342e+07
50031,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,53,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.321311e+07
50032,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,54,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.366974e+07
50033,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,55,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,3.412520e+07


## Reshape to load into tableau

In [26]:
# --- Read cost-benefit data ---
cb_data = results_all_pp_shifted.copy()

# Split 'variable' into components: name, sector, cb_type, item_1, item_2
# (Assumes exactly 5 colon-separated parts; if there are more colons inside the last field,
# they will be kept in item_2 thanks to n=4)
cb_chars = cb_data["variable"].astype(str).str.split(":", n=4, expand=True)
cb_chars.columns = ["name", "sector", "cb_type", "item_1", "item_2"]
cb_data = pd.concat([cb_data, cb_chars], axis=1)

In [27]:
cb_data.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value,name,sector,cb_type,item_1,item_2
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural


In [28]:
cb_data['strategy_code'].unique()

array(['PFLO:HBLE', 'PFLO:NDC_25_SCOE_ELEC', 'PFLO:NDC_25_UNCONDITIONAL',
       'PFLO:NDC_25_UNCONDITIONAL_CE'], dtype=object)

In [29]:
# Scale value from USD to billions (divide by 1e9)
if "value" in cb_data.columns:
    cb_data["value"] = cb_data["value"] / 1e9

# --- Remove "shifted" entries ---
# Remove rows where item_2 contains "shifted"
cb_data = cb_data[~cb_data["item_2"].astype(str).str.contains("shifted", na=False)]

# Remove any remaining rows where variable contains "shifted2"
cb_data = cb_data[~cb_data["variable"].astype(str).str.contains("shifted2", na=False)]

# --- Add Year column (Year = time_period + 2015) ---
cb_data["Year"] = cb_data["time_period"] + 2015

**This Section Needs Manual Tunnning**

In [30]:
cb_data.strategy_code.unique()

array(['PFLO:HBLE', 'PFLO:NDC_25_SCOE_ELEC', 'PFLO:NDC_25_UNCONDITIONAL',
       'PFLO:NDC_25_UNCONDITIONAL_CE'], dtype=object)

In [31]:
# Harmonize strategy name changes with a single mapping (no duplicate replacements)
strategy_name_map = {
    "PFLO:HBLE": "PFLO:HBLE",
    "PFLO:NDC_25_SCOE_ELEC": "PFLO:NDC_25_SCOE_ELEC",
    "PFLO:NDC_25_UNCONDITIONAL": "NDC_25_UNCONDITIONAL",
    "PFLO:NDC_25_UNCONDITIONAL_CE": "PFLO:NDC_25_UNCONDITIONAL_CE"
}

cb_data["strategy"] = (
    cb_data["strategy_code"].astype(str).map(strategy_name_map).fillna(cb_data["strategy_code"])
)
print(cb_data["strategy"].unique())


['PFLO:HBLE' 'PFLO:NDC_25_SCOE_ELEC' 'NDC_25_UNCONDITIONAL'
 'PFLO:NDC_25_UNCONDITIONAL_CE']


In [32]:
# --- Strategy ID mapping based on strategy_code ---
def map_strategy_id(code: str):
    if code == "BASE":
        return 0
    if code == "PFLO:HBLE":
        return 6004
    if code == "PFLO:NDC_25_SCOE_ELEC":
        return 6010
    if code == "PFLO:NDC_25_UNCONDITIONAL":
        return 6011
    if code == "PFLO:NDC_25_UNCONDITIONAL_CE":
        return 6012
    return code  # keep original if not matched

cb_data["strategy_id"] = cb_data["strategy_code"].astype(str).map(map_strategy_id)

In [33]:
ssp_data.primary_id.unique()

array([     0, 700070, 760076, 770077, 780078])

In [34]:
def map_primary_id(code: str):
    if code == "BASE":
        return 0
    if code == "PFLO:HBLE":
        return 700070
    if code == "PFLO:NDC_25_SCOE_ELEC":
        return 760076
    if code == "PFLO:NDC_25_UNCONDITIONAL":
        return 770077
    if code == "PFLO:NDC_25_UNCONDITIONAL_CE":
        return 780078
    return code  # keep original if not matched

cb_data["primary_id"] = cb_data["strategy_code"].astype(str).map(map_primary_id)


**Manual Tunning Ends Here**

In [35]:
# --- Create ids (variable:strategy_id) ---
cb_data["ids"] = cb_data["variable"].astype(str) + ":" + cb_data["strategy_id"].astype(str)

In [36]:
cb_data.head()

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value,name,sector,cb_type,item_1,item_2,Year,strategy,strategy_id,primary_id,ids
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2019,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2020,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2021,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2022,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.0,cb,wali,technical_cost,sanitation,unimp_rural,2023,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...


In [37]:
cb_data.strategy.unique()

array(['PFLO:HBLE', 'PFLO:NDC_25_SCOE_ELEC', 'NDC_25_UNCONDITIONAL',
       'PFLO:NDC_25_UNCONDITIONAL_CE'], dtype=object)

In [38]:
cb_data.strategy_code.unique()

array(['PFLO:HBLE', 'PFLO:NDC_25_SCOE_ELEC', 'PFLO:NDC_25_UNCONDITIONAL',
       'PFLO:NDC_25_UNCONDITIONAL_CE'], dtype=object)

In [39]:
# --- Read GDP and merge ---
# Keep only needed columns
gdp = ssp_data.loc[:, ["primary_id", "time_period", "gdp_mmm_usd"]]
gdp.head(100)

,primary_id,time_period,gdp_mmm_usd
0,0,4,96.640000
1,0,5,99.490000
2,0,6,103.010000
3,0,7,107.730000
4,0,8,113.483902
...,...,...,...
95,700070,47,885.672553
96,700070,48,903.386004
97,700070,49,921.453724
98,700070,50,939.882799


In [40]:
gdp[gdp['primary_id']== 760076]

,primary_id,time_period,gdp_mmm_usd
104,760076,4,96.640000
105,760076,5,99.490000
106,760076,6,103.010000
107,760076,7,107.730000
108,760076,8,113.483902
109,760076,9,120.406420
110,760076,10,128.353244
111,760076,11,137.787207
112,760076,12,150.256950
113,760076,13,165.012182


In [41]:
# Merge cb_data with gdp on primary_id and time_period
cb_data = cb_data.merge(gdp, on=["primary_id", "time_period"], how="left")
cb_data

,strategy_code,future_id,region,time_period,difference_variable,variable_value_baseline,variable_value_pathway,difference_value,variable,value,...,sector,cb_type,item_1,item_2,Year,strategy,strategy_id,primary_id,ids,gdp_mmm_usd
0,PFLO:HBLE,0,uganda,4,pop_unimproved_rural,1.932346e+07,1.932346e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2019,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...,96.640000
1,PFLO:HBLE,0,uganda,5,pop_unimproved_rural,1.951974e+07,1.951974e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2020,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...,99.490000
2,PFLO:HBLE,0,uganda,6,pop_unimproved_rural,1.896047e+07,1.896047e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2021,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...,103.010000
3,PFLO:HBLE,0,uganda,7,pop_unimproved_rural,1.837001e+07,1.837001e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2022,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...,107.730000
4,PFLO:HBLE,0,uganda,8,pop_unimproved_rural,1.774752e+07,1.774752e+07,0.0,cb:wali:technical_cost:sanitation:unimp_rural,0.000000,...,wali,technical_cost,sanitation,unimp_rural,2023,PFLO:HBLE,6004,700070,cb:wali:technical_cost:sanitation:unimp_rural:...,113.483902
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
44839,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,51,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.032291,...,scoe,indoor_air_pollution,None,None,2066,PFLO:NDC_25_UNCONDITIONAL_CE,6012,780078,cb:scoe:indoor_air_pollution:6012,958.680455
44840,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,52,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.032753,...,scoe,indoor_air_pollution,None,None,2067,PFLO:NDC_25_UNCONDITIONAL_CE,6012,780078,cb:scoe:indoor_air_pollution:6012,977.854064
44841,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,53,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.033213,...,scoe,indoor_air_pollution,None,None,2068,PFLO:NDC_25_UNCONDITIONAL_CE,6012,780078,cb:scoe:indoor_air_pollution:6012,997.411145
44842,PFLO:NDC_25_UNCONDITIONAL_CE,0,uganda,54,frac_scoe_heat_energy_residential_solid_biomass,NaN,NaN,NaN,cb:scoe:indoor_air_pollution,0.033670,...,scoe,indoor_air_pollution,None,None,2069,PFLO:NDC_25_UNCONDITIONAL_CE,6012,780078,cb:scoe:indoor_air_pollution:6012,1017.359368


In [42]:
96.640000/35.35

2.7338048090523337

In [43]:
cb_data[['time_period','gdp_mmm_usd']]

# 2019: 96.640000 #PPP
# 2019: 35.35 #USD
# 2.7
# https://data.worldbank.org/indicator/NY.GDP.MKTP.CD?locations=UG

,time_period,gdp_mmm_usd
0,4,96.640000
1,5,99.490000
2,6,103.010000
3,7,107.730000
4,8,113.483902
...,...,...
44839,51,958.680455
44840,52,977.854064
44841,53,997.411145
44842,54,1017.359368


In [44]:
cb_data.isna().sum()

strategy_code                 0
future_id                     0
region                        0
time_period                   0
difference_variable           0
variable_value_baseline     208
variable_value_pathway      364
difference_value            208
variable                      0
value                         0
name                          0
sector                        0
cb_type                       0
item_1                      208
item_2                     3340
Year                          0
strategy                      0
strategy_id                   0
primary_id                    0
ids                           0
gdp_mmm_usd                   0
dtype: int64

In [45]:
# Save the results
cb_data.to_csv(os.path.join(CB_OUTPUT, "cba_data_air_pollution.csv"), index = False)
